niworkflows/viz/plots.py

```diff
@@ -192,204 +190,106 @@ def plot_carpet(
-
         data = clean(data.T, t_r=tr, filter=False).T
-        v = (-2, 2)
+        vminmax = (np.percentile(data, 2), np.percentile(data, 98))
+
+    # Decimate number of time-series before clustering
+    n_dec = int((1.8 * data.shape[0]) // size[0])
+    if n_dec > 1:
+        segments = {
+            lab: idx[::n_dec] for lab, idx in segments.items()
+        }
 
     # Cluster segments (if argument enabled)
     if sort_rows:
         from scipy.cluster.hierarchy import linkage, dendrogram
         from sklearn.cluster import ward_tree
 
-        order = np.zeros(len(seg), dtype="uint32")
-        roi_start = 0
-        for i in np.unique(seg):
-            roi_mask = seg == i
-            roi = data[roi_mask]
-            if sort_rows.lower() == "linkage":
+        for seg_label, seg_idx in segments.items():
+            roi_data = data[seg_idx]
+            if isinstance(sort_rows, str) and sort_rows.lower() == "linkage":
                 linkage_matrix = linkage(
-                    roi, method="average", metric="euclidean", optimal_ordering=True
+                    roi_data, method="average", metric="euclidean", optimal_ordering=True
                 )
             else:
-                children, _, n_leaves, _, distances = ward_tree(roi, return_distance=True)
+                children, _, n_leaves, _, distances = ward_tree(roi_data, return_distance=True)
                 linkage_matrix = _ward_to_linkage(children, n_leaves, distances)
 
             dn = dendrogram(linkage_matrix, no_plot=True)
-            nreg = len(roi)
-            order[roi_start : roi_start + nreg] = np.argwhere(roi_mask).squeeze()[
-                np.array(dn["leaves"])
-            ]
-            roi_start += nreg
-    else:
-        order = seg.argsort()
+            # Override the ordering of the indices in this segment
+            segments[seg_label] = np.array(seg_idx)[np.array(dn["leaves"])]
 
     # If subplot is not defined
     if subplot is None:
         subplot = mgs.GridSpec(1, 1)[0]
 
+    # Length before decimation
+    n_trs = data.shape[-1] - drop_trs
+
     # Define nested GridSpec
-    wratios = [1, 100, 20]
     gs = mgs.GridSpecFromSubplotSpec(
-        1,
-        2 + int(legend),
-        subplot_spec=subplot,
-        width_ratios=wratios[: 2 + int(legend)],
-        wspace=0.0,
+        1, 2, subplot_spec=subplot, width_ratios=(1, 100), wspace=0.0,
     )
 
     # Segmentation colorbar
+    colors = np.hstack([[i + 1] * len(v) for i, v in enumerate(segments.values())])
     ax0 = plt.subplot(gs[0])
     ax0.set_yticks([])
     ax0.set_xticks([])
-    ax0.imshow(seg[order, np.newaxis], interpolation="none", aspect="auto", cmap=cmap)
+    ax0.imshow(
+        colors[:, np.newaxis],
+        interpolation="none",
+        aspect="auto",
+        cmap=cmap
+    )
 
     ax0.grid(False)
     ax0.spines["left"].set_visible(False)
     ax0.spines["bottom"].set_color("none")
     ax0.spines["bottom"].set_visible(False)
 
```

Comment:
I think this would be the good PR to add the legend and maybe the separation line between the carpetplot compartments
```suggestion
      import matplotlib.patches as mpatches
      leg = []
      for s, key in enumerate(segments.keys()):
          leg.append(mpatches.Patch(color=cmap.colors[s], label=key))
      plt.legend(handles=leg)
```

Reply:
I would leave this for a future PR. We might better just print the region name next to the colorbar.